Clumsy-Coder/pihole-dashboard
# 620 refactor GitHub action build workflow

# Summary

Refactor `.github/workflow/build.yml` to be more modular and use reusable workflow with matrix to build docker images

- build.yml
  - create job to extract next version and whether to publish docker image to ghcr
  - use matrix builds to run reusable workflow to build docker images with various tagnames
  - create job to verify job matrix has finished successfully. Used for required status checks
- reusable-docker-build.yml
  - create reusable github action to build docker images
  - takes input
    - `tags`: tagname of the docker image
    - `publish`: `true` to push to ghcr, `false` otherwise
- cleanup-docker-image
  - create workflow to remove development docker image after PR is closed and merged
  - ex: 
    - branch name `620` from `development` branch is created                                                   
    - committed some code                                                                                      
      - github action creates docker image with tagname `620`                                                  
    - created a pull request                                                                                   
    - after pull request closed and merged, this github action will remove the docker image 

# Description

- add reusable workflow `.github/workflows/reusable-docker-build.yml`
  - build docker image
  - push docker image
  - remove untagged docker images
- `.github/workflows/build.yml`
  - job `prepare-docker-build`
    - extract the next version
    - extract whether to publish no version or not
  - job `docker-build-image`
    - call reusable workflow `.github/workflows/reusable-docker-build.yml`
    - use matrix to build multiple docker images
      - version: latest
        - push docker image: only when PUBLISH is `true`
      - version: NEXT_VERSION
        - push docker image: only when PUBLISH is `true`
      - version: nightly
        - push docker image: only when on `development` branch
      - version: branch name ( git refname )
        - push docker image: only when not on branch `development` and `master`
  - job `matrix-results`
    - check if matrix job builds has succeeded
      - if any of them has failed return `exit 1`. this will signal github action a job has failed.
  - remove unused code
    - remove job `docker-build`
- `.github/workflows/cleanup-docker-image.yml`
  - run on Pull Requests that are closed and merged
  - fetch metadata of all docker images published on ghcr
  - extract version ID that uses docker image tagname of the development branch
    - ex:
      - create branch `620` from `development` branch
      - commit code
      - github action creates docker image with tagname `620`
      - create pull request of branch `620` to `development` branch
      - merge branch and close pull request
      - github action will find version ID of docker image with tagname `620`
      - remove docker image
  - remove docker image using version ID
- enforce required status check for job `matrix-results` to succeed

# Related Issue




#620

# Motivation and Context

- reduces the size of `.github/workflows/build.yml`
- use modularity to build docker image
- uses matrix to build docker image. less repeating code
- create and remove docker images used for development. for testing purposes

# How Has This Been Tested?




# Screenshots (if appropriate)

# Types of changes

- [x] **Style**: (fix code formatting issues, whitespace, missing semicolon, rename variable etc)
- [x] **Chore**: (add changes that are not related to source code. Ex: changing eslint config)
- [ ] **Code refactor**: (change working code without regressing features)
- [ ] **Test**: (adding/updating tests)
- [ ] **Build**: (changes that affect the build system. Ex: npm)
- [x] **CI**: (changes to the CI/CD configs and scripts)

## Does this Pull Request introduce a breaking change?


# Checklist


- [x] The commit message follows **conventional commits** guidelines: https://www.conventionalcommits.org/en/v1.0.0/
